language/libra-tools/writeset-transaction-generator/src/writeset_builder.rs

```diff
@@ -0,0 +1,180 @@
+                &mut CostStrategy::system(&ZERO_COST_SCHEDULE, GasUnits::new(100_000_000)),
+                &NoContextLog::new(),
+            )
+            .unwrap_or_else(|e| {
+                panic!(
+                    "Error calling {}.{}: {}",
+                    module_name,
+                    function_name,
+                    e.into_vm_status()
+                )
+            })
+    }
+
+    pub fn exec_script(&mut self, sender: AccountAddress, script: &Script) {
+        self.0
+            .execute_script(
+                script.code().to_vec(),
+                script.ty_args().to_vec(),
+                convert_txn_args(script.args()),
+                vec![sender],
+                &mut CostStrategy::system(&ZERO_COST_SCHEDULE, GasUnits::new(100_000_000)),
+                &NoContextLog::new(),
+            )
+            .unwrap()
+    }
+
+    fn disable_reconfiguration(&mut self) {
+        self.exec_func(
+            libra_root_address(),
+            "LibraConfig",
+            "disable_reconfiguration",
+            vec![],
+            vec![Value::transaction_argument_signer_reference(
+                libra_root_address(),
+            )],
+        )
+    }
+
+    fn enable_reconfiguration(&mut self) {
+        self.exec_func(
+            libra_root_address(),
+            "LibraConfig",
+            "enable_reconfiguration",
+            vec![],
+            vec![Value::transaction_argument_signer_reference(
+                libra_root_address(),
+            )],
+        )
+    }
+    pub fn set_libra_version(&mut self, version: Version) {
+        self.exec_func(
+            libra_root_address(),
+            "LibraVersion",
+            "set",
+            vec![],
+            vec![
+                Value::transaction_argument_signer_reference(libra_root_address()),
+                Value::u64(version),
+            ],
+        )
+    }
+}
+
+/// Convert the transaction arguments into Move values.
+fn convert_txn_args(args: &[TransactionArgument]) -> Vec<Value> {
+    args.iter()
+        .map(|arg| match arg {
+            TransactionArgument::U8(i) => Value::u8(*i),
+            TransactionArgument::U64(i) => Value::u64(*i),
+            TransactionArgument::U128(i) => Value::u128(*i),
+            TransactionArgument::Address(a) => Value::address(*a),
+            TransactionArgument::Bool(b) => Value::bool(*b),
+            TransactionArgument::U8Vector(v) => Value::vector_u8(v.clone()),
+        })
+        .collect()
+}
+
+fn move_module_changes(modules: &[CompiledModule]) -> MoveChanges {
+    let mut shadow_changeset = MoveChanges::new();
+    for module in modules.iter() {
+        let id = module.self_id();
+        let mut module_bytes = vec![];
+        module.serialize(&mut module_bytes).unwrap();
+        shadow_changeset.publish_or_overwrite_module(id, module_bytes);
+    }
+    shadow_changeset
+}
+
+pub fn build_changeset<S: StateView, F>(
+    state_view: &S,
+    procedure: F,
```

Comment:
Would it be possible to make this a `CompiledScript,  the name of a source file that will be compiled against` modules`, or a fully qualified procedure name? A closure feels like a bit of a blank check given how structured the uses of this are (e.g., typically calling one Move function).

My main concern here is transparency--it would be nice to publish a Move-only artifact on GitHub that shows how the WriteSet was produced. But this closure will disappear forever after `build_changeset` is run.

Reply:
There are two main issues here:
1. `CompiledScript` cannot invoke private functions in a module
2. `CompiledScript` cannot call as arbitrary signer.

Those are the two major drawbacks that already occurs in the genesis generation process that prevents us from using pure Move script to setup the states.